WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewModelTests.swift

```diff
@@ -0,0 +1,163 @@
+
+    }
+
+    func testHandlingTheOriginalDownloadableFilesIsAlwaysValidAndUnique() {
+
+    }
+
+    func testHandlingAValidDownloadableFilesAddsWithSuccess() {
+
+    }
+
+    func testHandlingAValidExistingDownloadableFileRemovesWithSuccess() {
+
+    }
+
+    // MARK: - `handleDownloadLimitChange`
+
+    func testHandlingAValidDownloadLimitUpdatesWithSuccess() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        viewModel.handleDownloadLimitChange(5)
+
+        // Assert
+        XCTAssertEqual(viewModel.downloadLimit, 5)
+    }
+
+    // MARK: - `handleDownloadExpiryChange`
+
+    func testHandlingAValidDownloadExpiryUpdatesWithSuccess() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        viewModel.handleDownloadExpiryChange(5)
+
+        // Assert
+        XCTAssertEqual(viewModel.downloadExpiry, 5)
+    }
+
+    // MARK: - `hasUnsavedChanges`
+
+    func testViewModelHasUnsavedChangesAfterUpdatingDownloadFileLimit() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        viewModel.handleDownloadLimitChange(5)
+
+        // Assert
+        XCTAssertTrue(viewModel.hasUnsavedChanges())
+
+    }
+
+    func testViewModelHasUnsavedChangesAfterUpdatingDownloadFileExpiry() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        viewModel.handleDownloadExpiryChange(5)
+
+        // Assert
+        XCTAssertTrue(viewModel.hasUnsavedChanges())
+    }
+
+    func testViewModelHasUnsavedChangesAfterUpdatingDownloadFilesOrder() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        guard let firstFile = viewModel.remove(at: 0) else {
+            XCTFail()
+            return
+        }
+        viewModel.insert(firstFile, at: 1)
+
+        // Assert
+        XCTAssertTrue(viewModel.hasUnsavedChanges())
+    }
+
+    func testViewModelHasUnsavedChangesAfterRemovingDownloadFileIndividually() {
+
+    }
+
+    func testViewModelHasUnsavedChangesAfterAddingDownloadFileIndividually() {
+
+    }
+
+    func testViewModelHasUnsavedChangesAfterUpdatingSingleDownloadFileIndividually() {
+
+    }
```

Comment:
Like above

Reply:
Those are intentionally left empty. Those are related to adding/removing downloadable files. But the functionality hasn't been implemented yet. Which is due to be implemented with #2758 and #2759 . So I left those empty for now, to implement with the respective tickets. But I think It will be better to remove them and add back again with those tickets.